File: src/test/java/org/proofit/premiumcalc/helper/PolicyBuilder.java
package org.proofit.premiumcalc.helper;

import org.proofit.premiumcalc.model.*;

import java.math.BigDecimal;
import java.util.ArrayList;
import java.util.List;
import java.util.function.Function;

public class PolicyBuilder {

    public static PolicyBuilder createPolicy() {
        return new PolicyBuilder();
    }

    private List<PolicyObject> policyObjects = new ArrayList<>();
    private int policyObjectCounter = 0;

    private PolicyBuilder() {
    }

    public PolicyBuilder withObject(Function<PolicyObjectBuilder, PolicyObjectBuilder> f) {
        policyObjects.add(f.apply(new PolicyObjectBuilder()).build(++policyObjectCounter));
        return this;
    }

    public Policy build() {
        return new Policy("Policy", PolicyStatus.REGISTERED, policyObjects);
    }

    public static class PolicyObjectBuilder {

        private List<PolicySubObject> subObjects = new ArrayList<>();

        private int subObjectCounter = 0;

        public PolicyObjectBuilder withSubObject(String sumInsured, RiskType riskType) {
            return withSubObject(new BigDecimal(sumInsured), riskType);
        }

        public PolicyObjectBuilder withSubObject(BigDecimal sumInsured, RiskType riskType) {
            subObjects.add(new PolicySubObject("SubObject_" + ++subObjectCounter, sumInsured, riskType));
            return this;
        }

        private PolicyObject build(int policyObjectNumber) {
            return new PolicyObject("PolicyObject_" + policyObjectNumber, subObjects);
        }
    }
}

File: src/test/kotlin/org/proofit/premiumcalc/helper/PolicyBuilder.kt
package org.proofit.premiumcalc.helper

import org.proofit.premiumcalc.model.Policy
import org.proofit.premiumcalc.model.PolicyObject
import org.proofit.premiumcalc.model.PolicyStatus
import org.proofit.premiumcalc.model.PolicySubObject
import org.proofit.premiumcalc.model.RiskType
import java.math.BigDecimal
import java.util.function.Function


class PolicyBuilder private constructor() {
    private val policyObjects: MutableList<PolicyObject> = mutableListOf()
    private var policyObjectCounter = 0

    fun withObject(f: Function<PolicyObjectBuilder, PolicyObjectBuilder>): PolicyBuilder {
        policyObjects.add(f.apply(PolicyObjectBuilder()).build(++policyObjectCounter))
        return this
    }

    fun build(): Policy {
        return Policy("Policy", PolicyStatus.REGISTERED, policyObjects)
    }

    class PolicyObjectBuilder {
        private val subObjects: MutableList<PolicySubObject> = mutableListOf()
        private var subObjectCounter = 0
        fun withSubObject(sumInsured: String, riskType: RiskType): PolicyObjectBuilder {
            return withSubObject(BigDecimal(sumInsured), riskType)
        }

        fun withSubObject(sumInsured: BigDecimal, riskType: RiskType): PolicyObjectBuilder {
            subObjects.add(PolicySubObject("SubObject_" + ++subObjectCounter, sumInsured, riskType))
            return this
        }

        internal fun build(policyObjectNumber: Int): PolicyObject {
            return PolicyObject("PolicyObject_$policyObjectNumber", subObjects)
        }
    }

    companion object {
        fun createPolicy(): PolicyBuilder {
            return PolicyBuilder()
        }
    }
}
